Mathlib/SetTheory/Nimber/Field.lean

```diff
@@ -0,0 +1,231 @@
+  rw [← one_add_one_eq_two, add_self]
+
+private theorem two_zsmul (x : Nimber) : (2 : ℤ) • x = 0 := by
+  rw [_root_.two_zsmul]
+  exact add_self x
+
+private theorem add_eq_iff_eq_add : a + b = c ↔ a = c + b :=
+  sub_eq_iff_eq_add
+
+/-- Nimber multiplication is recursively defined so that `a * b` is the smallest nimber not equal to
+`a' * b + a * b' + a' * b'` for `a' < a` and `b' < b`. -/
+-- We write the binders like this so that the termination checker works.
+protected def mul (a b : Nimber.{u}) : Nimber.{u} :=
+  sInf {x | ∃ a', ∃ (_ : a' < a), ∃ b', ∃ (_ : b' < b),
+    Nimber.mul a' b + Nimber.mul a b' + Nimber.mul a' b' = x}ᶜ
+termination_by (a, b)
+
+instance : Mul Nimber :=
+  ⟨Nimber.mul⟩
+
+theorem mul_def (a b : Nimber) :
+    a * b = sInf {x | ∃ a' < a, ∃ b' < b, a' * b + a * b' + a' * b' = x}ᶜ := by
+  change Nimber.mul a b = _
+  rw [Nimber.mul]
+  simp_rw [exists_prop]
+  rfl
+
+/-- The set in the definition of `Nimber.mul` is nonempty. -/
+private theorem mul_nonempty (a b : Nimber.{u}) :
+    {x | ∃ a' < a, ∃ b' < b, a' * b + a * b' + a' * b' = x}ᶜ.Nonempty := by
+  convert nonempty_of_not_bddAbove <| not_bddAbove_compl_of_small
+    ((fun x ↦ x.1 * b + a * x.2 + x.1 * x.2) '' Set.Iio a ×ˢ Set.Iio b)
+  ext
+  simp_rw [Set.mem_setOf_eq, Set.mem_image, Set.mem_prod, Set.mem_Iio, Prod.exists]
+  tauto
+
+theorem exists_of_lt_mul (h : c < a * b) : ∃ a' < a, ∃ b' < b, a' * b + a * b' + a' * b' = c := by
+  rw [mul_def] at h
+  have := not_mem_of_lt_csInf' h
+  rwa [Set.not_mem_compl_iff] at this
+
+theorem mul_le_of_forall_ne (h : ∀ a' < a, ∀ b' < b, a' * b + a * b' + a' * b' ≠ c) :
+    a * b ≤ c := by
+  by_contra! h'
+  have := exists_of_lt_mul h'
+  tauto
+
+instance : MulZeroClass Nimber where
+  mul_zero a := by
+    rw [← Nimber.le_zero]
+    exact mul_le_of_forall_ne fun _ _ _ h ↦ (Nimber.not_lt_zero _ h).elim
+  zero_mul a := by
+    rw [← Nimber.le_zero]
+    exact mul_le_of_forall_ne fun _ h ↦ (Nimber.not_lt_zero _ h).elim
+
+private theorem mul_ne_of_lt : ∀ a' < a, ∀ b' < b, a' * b + a * b' + a' * b' ≠ a * b := by
+  have H := csInf_mem (mul_nonempty a b)
+  rw [← mul_def] at H
+  simpa using H
+
+instance : NoZeroDivisors Nimber where
+  eq_zero_or_eq_zero_of_mul_eq_zero {a b} h := by
+    by_contra! hab
+    iterate 2 rw [← Nimber.pos_iff_ne_zero] at hab
+    apply (mul_ne_of_lt _ hab.1 _ hab.2).symm
+    simpa only [zero_add, mul_zero, zero_mul]
+
+protected theorem mul_comm (a b : Nimber) : a * b = b * a := by
+  apply le_antisymm <;>
+    apply mul_le_of_forall_ne <;>
+    intro x hx y hy
+  on_goal 1 => rw [add_comm (x * _), Nimber.mul_comm a, Nimber.mul_comm x, Nimber.mul_comm x]
+  on_goal 2 => rw [add_comm (x * _), ← Nimber.mul_comm y, ← Nimber.mul_comm a, ← Nimber.mul_comm y]
+  all_goals exact mul_ne_of_lt y hy x hx
+termination_by (a, b)
+
+protected theorem mul_add (a b c : Nimber) : a * (b + c) = a * b + a * c := by
+  apply le_antisymm
+  · apply mul_le_of_forall_ne
+    intro a' ha x hx
```

Comment:
nit: would be tempted to combine to
```suggestion
  · refine mul_le_of_forall_ne fun a' ha x hx => ?_
```
and elsewhere in this PR.

Reply:
I've gone ahead and done this.